src/Clash/Cores/Etherbone/WishboneMaster.hs

```diff
@@ -0,0 +1,146 @@
+  , selWidth ~ ByteSize dat
+  , Show dat
+  )
+  => WishboneMasterState dat
+  -> ( Bool
+     , ( Df.Data (WishboneOperation addrWidth selWidth dat)
+      , (Ack, WishboneS2M dat, ())
+      )
+     )
+  -> ( WishboneMasterState dat
+     , ( Ack
+       , ( Df.Data (WishboneResult dat)
+         , WishboneM2S addrWidth selWidth dat
+         , Maybe Bit)
+       )
+     )
+-- This operation is for another @AddressSpace@.
+wishboneMasterT state (rst, (Df.Data WishboneOperation{_opAddrSpace=ConfigAddressSpace}, _))
+  = (state, (Ack $ not rst, (Df.NoData, wbEmpty, Nothing)))
+    where
+      wbEmpty = emptyWishboneM2S
+wishboneMasterT state (rst, (iFwd, (Ack oBwd, wbBwd, _)))
+  = (nextState, (Ack iBwd, (oFwd, wbFwd, errBit)))
+  where
+    nextState = fsm state iFwd oBwd
+
+    wbErr = err wbBwd || retry wbBwd
+    wbTerm = acknowledge wbBwd || wbErr
+
+    oFwd = case state of
+      WaitForAck _ dat eor eop -> Df.Data $ WishboneResult dat eor eop
+      _                        -> Df.NoData
+
+    iBwd = not rst && case state of
+      WaitForOp _  -> False
+      Busy         -> False
+      WaitForAck{} -> oBwd
+
+    errBit = case (state, nextState) of
+      (Busy, WaitForAck{}) -> Just $ boolToBit wbErr
+      _                    -> Nothing
+
+    -- The wishbone bus receives the incoming operation in the transition from
+    -- @WaitForOp@ to @Busy@.
+    wbFwd = case (state, iFwd) of
+      (WaitForOp c, Df.NoData) -> wbEmpty   { strobe=False, busCycle=c }
+      (WaitForOp _, Df.Data i) -> (wbPkt i) { strobe=True,  busCycle=True }
+      (Busy, Df.NoData)        -> error "No input data in Busy state, this should be impossible!"
+      (Busy, Df.Data i)        -> (wbPkt i) { strobe=True,  busCycle=True }
+      (WaitForAck c _ _ _, _)  -> wbEmpty   { strobe=False, busCycle=c }
+    wbPkt WishboneOperation{..} = WishboneM2S
+      { addr                = _opAddr
+      , writeData           = fromMaybe undefined _opDat
+      , busSelect           = _opSel
+      , lock                = False
+      , busCycle            = undefined
+      , strobe              = undefined
+      , writeEnable         = isJust _opDat
+      , cycleTypeIdentifier = Classic
+      , burstTypeExtension  = LinearBurst
+      }
+    wbEmpty = emptyWishboneM2S
+
+    fsm
+      :: WishboneMasterState dat  -- state
+      -> Df.Data (WishboneOperation addrWidth selWidth dat)  -- iFwd
+      -> Bool                     -- oBwd
+      -> WishboneMasterState dat  -- nextState
+    fsm st@WaitForOp{} Df.NoData _ = st
+    fsm WaitForOp{} (Df.Data WishboneOperation{..}) _
+      | _opAbort  = WaitForOp False
+      | otherwise = Busy
+    fsm Busy{} Df.NoData _ = error "Sender did not keep Df channel constant!" 
+    fsm Busy{} (Df.Data x) _
+      | wbTerm    = WaitForAck (not $ _opDropCyc x) (dat $ _opDat x) (_opEOR x) (_opEOP x)
+      | otherwise = Busy
+      where
+        dat Nothing  = Just $ readData wbBwd
+        dat (Just _) = Nothing
```

Comment:
```suggestion
      | wbTerm    = WaitForAck (not $ _opDropCyc x) (readData wbBwd <$> _opDat x) (_opEOR x) (_opEOP x)
      | otherwise = Busy
```

Reply:
That's not right. If `_opDat` is `Just` it should return `Nothing`. And if `_opDat` is `Nothing` it should return a `Just` with the result from the WB bus.